File: src/main/java/org/cubexmc/metro/service/LineService.java
package org.cubexmc.metro.service;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.NamespacedKey;
import org.bukkit.util.Vector;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.manager.StopManager;
import org.cubexmc.metro.model.EntityModelController;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.model.Stop;
import org.cubexmc.metro.service.strategy.GlobalDispatchStrategy;
import org.cubexmc.metro.service.strategy.LocalDispatchStrategy;
import org.cubexmc.metro.train.TrainInstance;
import org.cubexmc.metro.util.LineTopologyUtil;
import org.cubexmc.metro.util.SchedulerUtil;

public class LineService {

    private final Metro plugin;
    private final LineServiceManager manager;
    private final String lineId;
    private final DispatchStrategy dispatchStrategy;
    private final TrainSpawner spawner;
    private final List<TrainInstance> activeTrains = new ArrayList<>();
    private final Set<TrainInstance> pendingRemoval = new LinkedHashSet<>();

    private int headwaySeconds;
    private int dwellTicks;
    private int trainCars;
    private long lastDepartureTick = -1L;
    private long totalTicks = 0L;
    private long totalTrainUpdates = 0L;
    private long totalSpawns = 0L;

    public LineService(Metro plugin, LineServiceManager manager, String lineId, int headwaySeconds, int dwellTicks,
            int trainCars, DispatchStrategy dispatchStrategy) {
        this.plugin = plugin;
        this.manager = manager;
        this.lineId = lineId;
        this.headwaySeconds = headwaySeconds;
        this.dwellTicks = dwellTicks;
        this.trainCars = Math.max(1, trainCars);
        this.dispatchStrategy = dispatchStrategy;
        this.spawner = new TrainSpawner(plugin, this);
    }

    public void tick() {
        totalTicks++;
        long currentTick = SchedulerUtil.getCurrentTick();
        if (dispatchStrategy != null) {
            dispatchStrategy.tick(this, currentTick);
        }
        updateActiveTrains(currentTick);
    }

    public void shutdown() {
        for (TrainInstance train : activeTrains) {
            queueTrainRemoval(train);
        }
        flushPendingTrainRemoval();
    }

    public void refreshStops() {
        Line line = getLine();
        if (line == null) return;
        List<String> newStops = line.getOrderedStopIds();

        long currentTick = SchedulerUtil.getCurrentTick();
        if (dispatchStrategy != null) {
            dispatchStrategy.refreshTopology(this, newStops, currentTick);
        }

        if (activeTrains.isEmpty()) return;

        LocalDispatchStrategy localDispatch = dispatchStrategy instanceof LocalDispatchStrategy local
                ? local : null;
        int retiredTrains = 0;
        int releasedVirtualTrains = 0;
        for (TrainInstance train : new ArrayList<>(activeTrains)) {
            if (localDispatch != null && train.getVirtualTrainId() != null) {
                localDispatch.releaseMaterializedTrain(train.getVirtualTrainId());
                releasedVirtualTrains++;
            }
            train.finishImmediately();
            queueTrainRemoval(train);
            retiredTrains++;
        }
        flushPendingTrainRemoval();

        plugin.getLogger().info("Refreshed topology for line " + lineId
                + "; retired " + retiredTrains + " active train(s)"
                + (releasedVirtualTrains > 0 ? " and released " + releasedVirtualTrains + " virtual train(s)" : "")
                + " to prevent stop-index desync.");
    }

    // ===== Delegated train management for TrainSpawner =====

    /**
     * Register a newly spawned train. Called by {@link TrainSpawner}.
     */
    public void addTrain(TrainInstance train, long currentTick) {
        activeTrains.add(train);
        totalSpawns++;
        manager.registerTrain(train);
    }

    // ===== Spawn methods (delegated to TrainSpawner) =====

    public void spawnTrain(long currentTick) {
        spawner.spawnTrainAtFirstStop(currentTick);
    }

    public TrainInstance spawnTrainForVirtual(long currentTick, int fromStopIndex, int toStopIndex,
            double progress, UUID virtualTrainId, String targetStopId) {
        return spawner.spawnTrainForVirtual(currentTick, fromStopIndex, toStopIndex,
                progress, virtualTrainId, targetStopId);
    }

    public void requestStop(String stopId, long currentTick) {
        if (dispatchStrategy != null) {
            dispatchStrategy.requestStop(this, stopId, currentTick);
        }
    }

    // ===== Train lifecycle =====

    private void updateActiveTrains(long currentTick) {
        Iterator<TrainInstance> iterator = activeTrains.iterator();
        while (iterator.hasNext()) {
            TrainInstance train = iterator.next();
            totalTrainUpdates++;
            train.update(currentTick);
            if (train.isFinished()) {
                queueTrainRemoval(train);
            }
        }
        flushPendingTrainRemoval();
    }

    public void handleTrainDerail(TrainInstance train) {
        if (train == null) return;
        train.finishImmediately();
        queueTrainRemoval(train);
    }

    public void virtualizeBackToPool(TrainInstance train, long currentTick) {
        if (dispatchStrategy instanceof LocalDispatchStrategy localDispatch) {
            localDispatch.returnTrainToPool(train, currentTick);
        }
    }

    private void queueTrainRemoval(TrainInstance train) {
        if (train != null) {
            pendingRemoval.add(train);
        }
    }

    private void flushPendingTrainRemoval() {
        if (pendingRemoval.isEmpty()) return;
        long currentTick = SchedulerUtil.getCurrentTick();
        for (TrainInstance train : pendingRemoval) {
            restoreVirtualAuthorityIfNeeded(train, currentTick);
            manager.unregisterTrain(train);
            train.cleanup();
            activeTrains.remove(train);
        }
        pendingRemoval.clear();
    }

    private void restoreVirtualAuthorityIfNeeded(TrainInstance train, long currentTick) {
        if (!(dispatchStrategy instanceof LocalDispatchStrategy localDispatch)) return;
        if (train == null || train.getVirtualTrainId() == null) return;
        if (!localDispatch.isMaterializedTrain(train.getVirtualTrainId())) return;
        localDispatch.returnTrainToPool(train, currentTick);
    }

    // ===== ETA / Scheduling =====

    public int estimateNextEtaSeconds(long currentTick) {
        int headwayTicks = getHeadwayTicks();
        if (lastDepartureTick < 0) {
            long mod = currentTick % headwayTicks;
            long etaTicks = (mod == 0) ? headwayTicks : (headwayTicks - mod);
            return (int) Math.ceil(etaTicks / 20.0);
        }
        long elapsed = currentTick - lastDepartureTick;
        if (elapsed < 0) elapsed = 0;
        long remaining = headwayTicks - (elapsed % headwayTicks);
        if (remaining == headwayTicks) remaining = 0;
        return (int) Math.ceil(remaining / 20.0);
    }

    public int estimateNextEtaSeconds(long currentTick, String stopId) {
        Integer virtualEta = estimateVirtualEtaSeconds(stopId);
        if (virtualEta != null) return virtualEta;

        double departureEtaSeconds = estimateNextEtaSeconds(currentTick);
        Line line = getLine();
        if (line == null || stopId == null || stopId.isEmpty()) {
            return (int) Math.ceil(departureEtaSeconds);
        }

        List<String> stopIds = line.getOrderedStopIds();
        double etaSeconds = ServiceEtaCalculator.estimateScheduledEtaSeconds(
                stopIds, stopId, dwellTicks, departureEtaSeconds,
                (fromStopId, toStopId) -> plugin.getTravelTimeEstimator().estimateSeconds(lineId, fromStopId, toStopId));
        return (int) Math.ceil(etaSeconds);
    }

    private Integer estimateVirtualEtaSeconds(String stopId) {
        if (!(dispatchStrategy instanceof LocalDispatchStrategy localDispatch)) return null;
        if (stopId == null || stopId.isEmpty()) return null;

        Line line = getLine();
        org.cubexmc.metro.service.virtual.VirtualTrainPool pool = localDispatch.getPool();
        if (line == null || pool == null) return null;

        List<String> stopIds = line.getOrderedStopIds();
        int stopIndex = stopIds.indexOf(stopId);
        if (stopIndex < 0) return null;

        org.cubexmc.metro.service.virtual.VirtualTrain candidate = pool.findBestCandidateForStop(
                stopId, stopIds, Double.POSITIVE_INFINITY,
                plugin.getTravelTimeEstimator(), SchedulerUtil.getCurrentTick());
        if (candidate == null) return null;

        double etaSeconds = candidate.estimateEtaToStop(stopIndex, plugin.getTravelTimeEstimator(), SchedulerUtil.getCurrentTick());
        if (!Double.isFinite(etaSeconds)) return null;
        return Math.max(0, (int) Math.round(etaSeconds));
    }

    public boolean isDepartureWindow(long currentTick) {
        return lastDepartureTick < 0 || currentTick - lastDepartureTick >= getHeadwayTicks();
    }

    public void markDeparture(long currentTick) {
        lastDepartureTick = currentTick;
    }

    public boolean tryMarkDeparture(long currentTick) {
        if (isDepartureWindow(currentTick)) {
            markDeparture(currentTick);
            return true;
        }
        return false;
    }

    public void markVirtualizedTrain(long currentTick) {
        if (isGlobalMode()) return;
        lastDepartureTick = Math.max(0L, currentTick - getHeadwayTicks());
    }

    // ===== Travel direction =====

    public Vector computeTravelDirection(Stop fromStop, Stop toStop) {
        if (fromStop == null) return new Vector(0, 0, 1);
        Vector direction = org.cubexmc.metro.util.LocationUtil.vectorFromYaw(fromStop.getLaunchYaw());
        return (direction != null && direction.lengthSquared() > 0) ? direction.normalize() : new Vector(0, 0, 1);
    }

    // ===== Getters / Accessors =====

    public String getLineId() { return lineId; }
    public int getHeadwaySeconds() { return headwaySeconds; }
    public int getHeadwayTicks() { return Math.max(20, headwaySeconds * 20); }
    public int getDwellTicks() { return dwellTicks; }
    public int getTrainCars() { return trainCars; }
    public Metro getPlugin() { return plugin; }
    public NamespacedKey getTrainKey() { return manager.getTrainKey(); }
    public BlockSectionManager getBlockSectionManager() { return manager.getBlockSectionManager(); }
    public StopManager getStopManager() { return plugin.getStopManager(); }
    public Line getLine() { return plugin.getLineManager().getLine(lineId); }
    public List<TrainInstance> getActiveTrains() { return new ArrayList<>(activeTrains); }

    public double getCartSpeed() {
        Line line = getLine();
        if (line != null) {
            double lineSpeed = line.getMaxSpeed();
            if (lineSpeed > 0) return lineSpeed;
        }
        return plugin.getCartSpeed();
    }

    public double getTrainSpacing() {
        Line line = getLine();
        String entityType = line != null ? line.getEntityType() : EntityModelController.MINECART_ENTITY_TYPE;
        if (line != null && !line.hasEntityTypeOverride()
                && plugin.getConfig().getBoolean("entity-model.enabled", false)
                && plugin.getEntityModelController() != null) {
            entityType = plugin.getEntityModelController().getDefaultEntityTypeRaw();
        }
        if (plugin.getEntityModelController() != null) {
            return plugin.getEntityModelController().getRecommendedSpacing(entityType, plugin.getTrainSpacing());
        }
        return EntityModelController.recommendedSpacing(entityType, plugin.getTrainSpacing());
    }
    public String buildSectionKey(String fromStopId, String toStopId) { return lineId + ":" + fromStopId + "->" + toStopId; }
    public boolean isGlobalMode() { return dispatchStrategy instanceof GlobalDispatchStrategy; }

    public boolean isLoopLine() {
        Line line = getLine();
        return line != null && LineTopologyUtil.isLoop(line.getOrderedStopIds());
    }

    public void setHeadwaySeconds(int headwaySeconds) { this.headwaySeconds = headwaySeconds; }
    public void setDwellTicks(int dwellTicks) { this.dwellTicks = dwellTicks; }
    public void setTrainCars(int trainCars) { this.trainCars = Math.max(1, trainCars); }

    public void refreshPhysicsEngines() {
        for (TrainInstance train : activeTrains) {
            train.refreshPhysicsEngine();
        }
    }

    public void refreshEntityModels() {
        for (TrainInstance train : activeTrains) {
            train.refreshEntityModels();
        }
    }

    public long getTotalTicks() { return totalTicks; }
    public long getTotalTrainUpdates() { return totalTrainUpdates; }
    public long getTotalSpawns() { return totalSpawns; }
}

File: src/main/java/org/cubexmc/metro/service/LineService.kt
package org.cubexmc.metro.service

import org.bukkit.NamespacedKey
import org.bukkit.util.Vector
import org.cubexmc.metro.Metro
import org.cubexmc.metro.manager.StopManager
import org.cubexmc.metro.model.EntityModelController
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.service.strategy.GlobalDispatchStrategy
import org.cubexmc.metro.service.strategy.LocalDispatchStrategy
import org.cubexmc.metro.train.TrainInstance
import org.cubexmc.metro.util.LineTopologyUtil
import org.cubexmc.metro.util.LocationUtil
import org.cubexmc.metro.util.SchedulerUtil
import java.util.LinkedHashSet
import java.util.UUID
import java.util.function.ToDoubleBiFunction
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.roundToInt

class LineService(
    val plugin: Metro,
    private val manager: LineServiceManager,
    val lineId: String,
    headwaySeconds: Int,
    dwellTicks: Int,
    trainCars: Int,
    private val dispatchStrategy: DispatchStrategy?,
) {
    private val spawner = TrainSpawner(plugin, this)
    private val activeTrainStorage = ArrayList<TrainInstance>()
    private val pendingRemoval: MutableSet<TrainInstance> = LinkedHashSet()

    var headwaySeconds = headwaySeconds
    var dwellTicks = dwellTicks
    var trainCars = max(1, trainCars)
        set(value) {
            field = max(1, value)
        }
    private var lastDepartureTick = -1L
    var totalTicks = 0L
        private set
    var totalTrainUpdates = 0L
        private set
    var totalSpawns = 0L
        private set

    val headwayTicks: Int
        get() = max(20, headwaySeconds * 20)
    val trainKey: NamespacedKey
        get() = manager.getTrainKey()
    val blockSectionManager: BlockSectionManager
        get() = manager.getBlockSectionManager()
    val stopManager: StopManager
        get() = plugin.stopManager
    val line: Line?
        get() = plugin.lineManager.getLine(lineId)
    val activeTrains: List<TrainInstance>
        get() = ArrayList(activeTrainStorage)
    val cartSpeed: Double
        get() {
            val currentLine = line
            if (currentLine != null) {
                val lineSpeed = currentLine.getMaxSpeed() ?: throw NullPointerException("line max speed")
                if (lineSpeed > 0) {
                    return lineSpeed
                }
            }
            return plugin.cartSpeed
        }
    val trainSpacing: Double
        get() {
            val currentLine = line
            var entityType: String? = if (currentLine != null) {
                lineEntityTypeOrNull(currentLine)
            } else {
                EntityModelController.MINECART_ENTITY_TYPE
            }
            val modelController = plugin.entityModelController
            if (currentLine != null && !currentLine.hasEntityTypeOverride() &&
                plugin.config.getBoolean("entity-model.enabled", false) && modelController != null
            ) {
                entityType = modelController.getDefaultEntityTypeRaw()
            }
            return if (modelController != null) {
                modelController.getRecommendedSpacing(entityType, plugin.trainSpacing)
            } else {
                EntityModelController.recommendedSpacing(entityType, plugin.trainSpacing)
            }
        }
    val isGlobalMode: Boolean
        get() = dispatchStrategy is GlobalDispatchStrategy
    val isLoopLine: Boolean
        get() {
            val currentLine = line ?: return false
            return LineTopologyUtil.isLoop(currentLine.orderedStopIds)
        }

    fun tick() {
        totalTicks++
        val currentTick = SchedulerUtil.getCurrentTick()
        dispatchStrategy?.tick(this, currentTick)
        updateActiveTrains(currentTick)
    }

    fun shutdown() {
        for (train in activeTrainStorage) {
            queueTrainRemoval(train)
        }
        flushPendingTrainRemoval()
    }

    fun refreshStops() {
        val line = line ?: return
        val newStops = line.orderedStopIds
        val currentTick = SchedulerUtil.getCurrentTick()
        dispatchStrategy?.refreshTopology(this, newStops, currentTick)
        if (activeTrainStorage.isEmpty()) {
            return
        }

        val localDispatch = dispatchStrategy as? LocalDispatchStrategy
        var retiredTrains = 0
        var releasedVirtualTrains = 0
        for (train in ArrayList(activeTrainStorage)) {
            val virtualTrainId = train.virtualTrainId
            if (localDispatch != null && virtualTrainId != null) {
                localDispatch.releaseMaterializedTrain(virtualTrainId)
                releasedVirtualTrains++
            }
            train.finishImmediately()
            queueTrainRemoval(train)
            retiredTrains++
        }
        flushPendingTrainRemoval()

        plugin.logger.info(
            "Refreshed topology for line $lineId; retired $retiredTrains active train(s)" +
                if (releasedVirtualTrains > 0) {
                    " and released $releasedVirtualTrains virtual train(s) to prevent stop-index desync."
                } else {
                    " to prevent stop-index desync."
                },
        )
    }

    fun addTrain(train: TrainInstance, currentTick: Long) {
        activeTrainStorage.add(train)
        totalSpawns++
        manager.registerTrain(train)
    }

    fun spawnTrain(currentTick: Long) {
        spawner.spawnTrainAtFirstStop(currentTick)
    }

    fun spawnTrainForVirtual(
        currentTick: Long,
        fromStopIndex: Int,
        toStopIndex: Int,
        progress: Double,
        virtualTrainId: UUID,
        targetStopId: String,
    ): TrainInstance? = spawner.spawnTrainForVirtual(
        currentTick,
        fromStopIndex,
        toStopIndex,
        progress,
        virtualTrainId,
        targetStopId,
    )

    fun requestStop(stopId: String, currentTick: Long) {
        dispatchStrategy?.requestStop(this, stopId, currentTick)
    }

    private fun updateActiveTrains(currentTick: Long) {
        for (train in activeTrainStorage) {
            totalTrainUpdates++
            train.update(currentTick)
            if (train.isFinished) {
                queueTrainRemoval(train)
            }
        }
        flushPendingTrainRemoval()
    }

    fun handleTrainDerail(train: TrainInstance?) {
        if (train == null) {
            return
        }
        train.finishImmediately()
        queueTrainRemoval(train)
    }

    fun virtualizeBackToPool(train: TrainInstance, currentTick: Long) {
        (dispatchStrategy as? LocalDispatchStrategy)?.returnTrainToPool(train, currentTick)
    }

    private fun queueTrainRemoval(train: TrainInstance?) {
        if (train != null) {
            pendingRemoval.add(train)
        }
    }

    private fun flushPendingTrainRemoval() {
        if (pendingRemoval.isEmpty()) {
            return
        }
        val currentTick = SchedulerUtil.getCurrentTick()
        for (train in pendingRemoval) {
            restoreVirtualAuthorityIfNeeded(train, currentTick)
            manager.unregisterTrain(train)
            train.cleanup()
            activeTrainStorage.remove(train)
        }
        pendingRemoval.clear()
    }

    private fun restoreVirtualAuthorityIfNeeded(train: TrainInstance?, currentTick: Long) {
        val localDispatch = dispatchStrategy as? LocalDispatchStrategy ?: return
        val virtualTrainId = train?.virtualTrainId ?: return
        if (!localDispatch.isMaterializedTrain(virtualTrainId)) {
            return
        }
        localDispatch.returnTrainToPool(train, currentTick)
    }

    fun estimateNextEtaSeconds(currentTick: Long): Int {
        val headwayTicks = headwayTicks
        if (lastDepartureTick < 0) {
            val mod = currentTick % headwayTicks
            val etaTicks = if (mod == 0L) headwayTicks.toLong() else headwayTicks - mod
            return ceil(etaTicks / 20.0).toInt()
        }
        var elapsed = currentTick - lastDepartureTick
        if (elapsed < 0) {
            elapsed = 0
        }
        var remaining = headwayTicks - elapsed % headwayTicks
        if (remaining == headwayTicks.toLong()) {
            remaining = 0
        }
        return ceil(remaining / 20.0).toInt()
    }

    fun estimateNextEtaSeconds(currentTick: Long, stopId: String?): Int {
        val virtualEta = estimateVirtualEtaSeconds(stopId)
        if (virtualEta != null) {
            return virtualEta
        }
        val departureEtaSeconds = estimateNextEtaSeconds(currentTick).toDouble()
        val line = line
        if (line == null || stopId.isNullOrEmpty()) {
            return ceil(departureEtaSeconds).toInt()
        }
        val etaSeconds = ServiceEtaCalculator.estimateScheduledEtaSeconds(
            line.orderedStopIds,
            stopId,
            dwellTicks,
            departureEtaSeconds,
            ToDoubleBiFunction { fromStopId, toStopId ->
                plugin.travelTimeEstimator.estimateSeconds(lineId, fromStopId, toStopId)
            },
        )
        return ceil(etaSeconds).toInt()
    }

    private fun estimateVirtualEtaSeconds(stopId: String?): Int? {
        val localDispatch = dispatchStrategy as? LocalDispatchStrategy ?: return null
        if (stopId.isNullOrEmpty()) {
            return null
        }
        val line = line ?: return null
        val pool = localDispatch.pool ?: return null
        val stopIds = line.orderedStopIds
        val stopIndex = stopIds.indexOf(stopId)
        if (stopIndex < 0) {
            return null
        }
        val candidate = pool.findBestCandidateForStop(
            stopId,
            stopIds,
            Double.POSITIVE_INFINITY,
            plugin.travelTimeEstimator,
            SchedulerUtil.getCurrentTick(),
        ) ?: return null
        val etaSeconds = candidate.estimateEtaToStop(
            stopIndex,
            plugin.travelTimeEstimator,
            SchedulerUtil.getCurrentTick(),
        )
        if (!etaSeconds.isFinite()) {
            return null
        }
        return max(0, etaSeconds.roundToInt())
    }

    fun isDepartureWindow(currentTick: Long): Boolean =
        lastDepartureTick < 0 || currentTick - lastDepartureTick >= headwayTicks

    fun markDeparture(currentTick: Long) {
        lastDepartureTick = currentTick
    }

    fun tryMarkDeparture(currentTick: Long): Boolean {
        if (!isDepartureWindow(currentTick)) {
            return false
        }
        markDeparture(currentTick)
        return true
    }

    fun markVirtualizedTrain(currentTick: Long) {
        if (!isGlobalMode) {
            lastDepartureTick = max(0L, currentTick - headwayTicks)
        }
    }

    fun computeTravelDirection(fromStop: Stop?, toStop: Stop?): Vector {
        if (fromStop == null) {
            return Vector(0, 0, 1)
        }
        val direction = vectorFromYawOrNull(fromStop.launchYaw)
        return if (direction != null && direction.lengthSquared() > 0) {
            direction.normalize()
        } else {
            Vector(0, 0, 1)
        }
    }

    fun buildSectionKey(fromStopId: String, toStopId: String): String =
        "$lineId:$fromStopId->$toStopId"

    private fun vectorFromYawOrNull(yaw: Float): Vector? = LocationUtil.vectorFromYaw(yaw)

    private fun lineEntityTypeOrNull(line: Line): String? = line.getEntityType()

    fun refreshPhysicsEngines() {
        for (train in activeTrainStorage) {
            train.refreshPhysicsEngine()
        }
    }

    fun refreshEntityModels() {
        for (train in activeTrainStorage) {
            train.refreshEntityModels()
        }
    }

}
